compiler/rustc_parse/src/parser/item.rs

```diff
@@ -449,6 +450,21 @@ impl<'a> Parser<'a> {
         if end.is_doc_comment() {
             err.span_label(end.span, "this doc comment doesn't document anything");
         }
+        if end.meta_kind().is_some() {
+            let pos = end.span.hi() + BytePos(1);
+            let semi_span = end.span.shrink_to_hi().with_hi(pos);
+            match self.sess.source_map().span_to_snippet(semi_span) {
+                Ok(snippet) if snippet.ends_with(';') => {
```

Comment:
Instead of fetching the source map and doing a textual comparison, could we check the current token to be a semicolon?